clang/lib/CodeGen/Targets/HLSLTargetInfo.h

```diff
@@ -0,0 +1,39 @@
+//===- HLSLTargetInfo.h ---------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "ABIInfoImpl.h"
+#include "TargetInfo.h"
+
+using namespace clang;
+using namespace clang::CodeGen;
+
+//===----------------------------------------------------------------------===//
+// Target codegen info implementation common between DirectX and SPIR/SPIR-V.
+//===----------------------------------------------------------------------===//
+
+class CommonHLSLTargetCodeGenInfo : public TargetCodeGenInfo {

```

Comment:
I worry a little bit about the composability of interposing this class between TargetCodeGenInfo and the DirectX and SPIRV targets. Imagine if we added a second language-specific codegeninfo - how would a backend be able to opt-in to supporting both?

I guess it would be better to make this a CRTP mixin, or to have `createHLSLBufferLayoutType` as a free function with a callback for `getHLSLLayoutType`.

Reply:
(in fact, maybe just a `StringRef Name` rather than having `getHLSLLayoutType` - `createHLSLBufferLayoutType` seems to depend pretty heavily on this returning a very specific type)

Reply:
See above. The layout calculations were move to a standalone `HLSLBufferLayoutBuilder` class.